### PR TITLE
Kafka-3375: Allows Watermark Generation in the Kafka Source.

### PR DESCRIPTION
**Pull request by kl0u**
In a nutshell this PR allows the generation of Watermarks in the Kafka source, based on a user-provided function. This is to face the situation where a Kafka source has multiple partitions per source task, and records become out of order before timestamps can be extracted and watermarks can be generated. 

This PR implements 2 new versions of the FlinkKafkaConsumer for each of the supported Kafka version (0.8 and 0.9), with each one taking an additional argument, which is: for the first a AssignerWithPeriodicWatermarks() timestamp extractor and for the second, a AssignerWithPunctuatedWatermarks().

### CODE REVIEW

File: flink-streaming-connectors/flink-connector-kafka-0.8/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer08WithPeriodicWM.java [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+		}
+		setNextWatermarkTimer();
+	}
+
+	private void setNextWatermarkTimer() {
+		long timeToNextWatermark = getTimeToNextWaternark();
+		runtime.registerTimer(timeToNextWatermark, this);
+	}
+
+	private long getTimeToNextWaternark() {
```
Comment (tillrohrmann) [inline flink-streaming-connectors/flink-connector-kafka-0.8/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer08WithPeriodicWM.java]: typo: Watermark

**Comment by mxm** [comment]
It looks like the change is API-breaking in terms of binary compatibility because it changes the base class of the `FlinkKafkaConsumer08` and `FlinkKafkaConsumer09`. I wonder if that is a problem.

**Comment by kl0u** [comment]
Thanks a lot @mxm !
If it is a problem, I can put the Watermark-related code in the FlinkConsumerBase class.
It is just that this way it is clear which code is responsible for which functionality.

**Comment by mxm** [comment]
@kl0u Has this been merged or are you going to open a new pr?

**Comment by kl0u** [comment]
No I opened a new one. It is this one https://github.com/apache/flink/pull/1839

**Pull request closed without merging**